core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/BlockAvailabilityLegacyAdapter.java

```diff
@@ -0,0 +1,188 @@
+            for (var unavailableSegment : unavailableSpace.get(blockWithOffset.blockID)) {
+                var trainInBlock = timeTrainInBlock(
+                        unavailableSegment,
+                        blockWithOffset,
+                        startOffset,
+                        envelope,
+                        startTime
+                );
+                if (trainInBlock == null)
+                    continue;
+                if (trainInBlock.start < unavailableSegment.timeEnd()
+                        && trainInBlock.end > unavailableSegment.timeStart()) {
+                    var blockMinimumDelay = unavailableSegment.timeEnd() - trainInBlock.start;
+                    if (blockMinimumDelay > minimumDelay) {
+                        minimumDelay = blockMinimumDelay;
+                        if (trainInBlock.start <= unavailableSegment.timeStart()) {
+                            // The train enters the block before it's unavailable: conflict at end location
+                            conflictOffset = blockWithOffset.pathOffset() + unavailableSegment.distanceEnd();
+                        } else {
+                            // The train enters the block when it's already unavailable: conflict at start location
+                            conflictOffset = blockWithOffset.pathOffset() + unavailableSegment.distanceStart();
+                        }
+                    }
+                }
+            }
+        }
+        if (minimumDelay == 0)
+            return null;
+        if (Double.isFinite(minimumDelay)) {
+            // We need to add delay, a recursive call is needed to detect new conflicts that appear with the added delay
+            var recursive = findMinimumDelay(blocks, startOffset, endOffset, envelope, startTime + minimumDelay);
+            if (recursive != null)
+                minimumDelay += recursive.duration;
+        }
+        var pathLength = blocks.stream()
+                .mapToLong(block -> blockInfra.getBlockLength(block.blockID))
+                .sum();
+        conflictOffset = Math.max(0, Math.min(pathLength, conflictOffset));
+        return new Unavailable(minimumDelay, conflictOffset);
+    }
+
+    /** Find the maximum amount of delay that can be added to the train without causing conflict.
+     * Cannot be called if the train is currently causing a conflict. */
+    Available findMaximumDelay(
+            List<BlockWithOffset> blocks,
+            long startOffset,
+            long endOffset,
+            EnvelopeTimeInterpolate envelope,
+            double startTime
+    ) {
+        double maximumDelay = Double.POSITIVE_INFINITY;
+        double timeOfNextOccupancy = Double.POSITIVE_INFINITY;
+        for (var blockWithOffset : iterateBlocksInRange(blocks, startOffset, endOffset)) {
+            for (var block : unavailableSpace.get(blockWithOffset.blockID)) {
+                var trainInBlock = timeTrainInBlock(
```

Comment:
We lose the time info here. It's longer, but I'd rather just have `var timeTrainInBlock`.